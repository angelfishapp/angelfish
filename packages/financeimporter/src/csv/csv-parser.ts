import type { ParsedData, Parser } from '../types/parser'
import type { Transaction } from '../types/transaction'
import { parseDate } from '../utils'
import type { CSVHeader, CSVTransactionMapper } from './csv-parser-types'

/**
 * Parser for CSV Files. Unlike other parsers, this parser requires a mapper to map the CSV data to the Transaction object.
 * You can use getHeaders() to get the headers in the CSV file and use it to create the mapper.
 */
export class CSVParser implements Parser {
  /**
   * Use this function to parse the CSV File String Data
   *
   * @param fileData    The file data as string to parse
   * @param mapper      The mapper to use to map the data to the Transaction object
   * @param delimiter   The delimiter to use to split the row. Defaults to ','
   * @param isDDMMYY    Whether the date format is DD/MM/YY (true) or MM/DD/YY (false). Defaults to true
   * @returns           A promise that resolves to a ParsedData object
   */
  async parse(fileData: string, mapper: CSVTransactionMapper): Promise<ParsedData> {
    const lines = fileData.split('\n')
    const rows = lines
      .filter((row) => row !== '') // Remove empty rows
      .map((row) => this.parseCSVRow(row, mapper.settings.csv_delimiter))
    const headers = rows[0]

    // Create header/key map
    const map = new Map()
    for (const [key, value] of Object.entries(mapper.fields)) {
      map.set(value, key)
    }

    // Iterate through each row and map the data to the Transaction object
    const transactions: Transaction[] = rows.slice(1).map((row) => {
      const transaction: any = {}
      row.forEach((value, index) => {
        const header = headers[index]
        const key = map.get(header)
        if (key) {
          if (key === 'date') {
            transaction[key] = parseDate(
              value,
              mapper.settings.date_format === 'DD MM YY' ||
                mapper.settings.date_format === 'DD MM YYYY',
            )
          } else if (key === 'amount') {
            transaction[key] = parseFloat(value)
          } else {
            transaction[key] = value
          }
        }
      })
      return transaction as Transaction
    })

    return { accounts: [], transactions }
  }

  /**
   * Get the CSV headers in the file as an array of strings
   *
   * @param fileData   The file data as string to parse
   * @param delimiter  The delimiter to use to split the row. Defaults to ','
   * @returns          An array of CSVHeader objects representing the headers in the file
   */
  getHeaders(fileData: string, delimiter: string = ','): CSVHeader[] {
    const lines = fileData.split('\n')

    // If there is only one line, return the headers from that line
    if (lines.length >= 1) {
      const headers = this.parseCSVRow(lines[0], delimiter)
      // For each header get up to 5 sample values from the column
      return headers.map((header, index) => {
        const samples = lines.slice(1, 6).map((row) => {
          return this.parseCSVRow(row, delimiter)[index]
        })
        return { header, samples }
      })
    }
  }

  /**
   * Parse an individual row in the CSV file. Provides support for quoted string values
   * in the CSV file.
   *
   * @param row         The full row as string to parse
   * @param delimiter   The delimiter to use to split the row.
   * @returns           An array of strings representing the column values in the row
   */
  private parseCSVRow(row: string, delimiter: string): string[] {
    const columns: string[] = []
    let currentCol = ''
    let insideQuotes = false

    for (let i = 0; i < row.length; i++) {
      const char = row[i]

      // Toggle the insideQuotes flag if we encounter a quotation mark
      if (char === '"') {
        insideQuotes = !insideQuotes
        continue
      }

      // Add characters to the current column if we're inside quotes
      if (insideQuotes) {
        currentCol += char
        continue
      }

      // If we encounter the delimiter and we're not inside quotes, we've reached the end of a column
      if (char === delimiter && !insideQuotes) {
        columns.push(currentCol.trim())
        currentCol = ''
      } else {
        currentCol += char
      }
    }

    // Add the last column
    columns.push(currentCol.trim())
    return columns
  }
}
