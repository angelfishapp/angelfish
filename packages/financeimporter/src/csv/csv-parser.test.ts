import { readFileSync } from 'fs'
import * as path from 'path'

import { CSVParser } from './csv-parser'

const TEST_FILES_DIR = path.resolve(__dirname, '__testfiles__')

describe('deserializeCSV()', () => {
  it('parse-headers-with-quotes', async () => {
    const csvString = 'Name, "Occupation, Not Separated", Age\nJohn, Software Developer, 30'
    const csv = new CSVParser()
    expect(csv.getHeaders(csvString)).toMatchSnapshot()
  })

  it('parse-csv-file-headers', async () => {
    const file = readFileSync(TEST_FILES_DIR + '/statement.csv').toString('utf-8')
    const csv = new CSVParser()

    expect(csv.getHeaders(file)).toMatchSnapshot()
  })

  it('parse-csv-file', async () => {
    const file = readFileSync(TEST_FILES_DIR + '/statement.csv').toString('utf-8')
    const csv = new CSVParser()

    expect(
      await csv.parse(file, {
        fields: {
          date: 'Posting Date',
          amount: 'Amount',
          name: 'Description',
          check_number: 'Check or Slip #',
        },
        settings: {
          csv_delimiter: ',',
          date_format: 'MM DD YYYY',
        },
      }),
    ).toMatchSnapshot()
  })
})
