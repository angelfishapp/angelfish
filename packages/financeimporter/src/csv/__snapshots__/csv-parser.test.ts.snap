// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deserializeCSV() > parse-csv-file 1`] = `
{
  "accounts": [],
  "transactions": [
    {
      "amount": -1400,
      "check_number": "",
      "date": 2023-08-21T00:00:00.000Z,
      "name": "PayPal",
    },
    {
      "amount": -22.7,
      "check_number": "",
      "date": 2023-08-21T00:00:00.000Z,
      "name": "Uber",
    },
    {
      "amount": -2332.34,
      "check_number": "",
      "date": 2023-08-21T00:00:00.000Z,
      "name": "CHASE CREDIT CRD AUTOPAY",
    },
    {
      "amount": -330,
      "check_number": "",
      "date": 2023-08-21T00:00:00.000Z,
      "name": "ATM WITHDRAWAL",
    },
    {
      "amount": -88.12,
      "check_number": "",
      "date": 2023-08-21T00:00:00.000Z,
      "name": "Macdonalds",
    },
    {
      "amount": 65,
      "check_number": "",
      "date": 2023-08-21T00:00:00.000Z,
      "name": "Zelle Payment to John Doe",
    },
    {
      "amount": 150,
      "check_number": "",
      "date": 2023-08-21T00:00:00.000Z,
      "name": "Zelle Payment to John Doe",
    },
    {
      "amount": 5000,
      "check_number": "",
      "date": 2023-08-18T00:00:00.000Z,
      "name": "ACME, INC.",
    },
    {
      "amount": 25.22,
      "check_number": "1",
      "date": 2023-08-16T00:00:00.000Z,
      "name": "REMOTE ONLINE DEPOSIT #          1",
    },
    {
      "amount": -500,
      "check_number": "",
      "date": 2023-08-14T00:00:00.000Z,
      "name": "Online Transfer to SAV",
    },
    {
      "amount": -86,
      "check_number": "",
      "date": 2023-08-11T00:00:00.000Z,
      "name": "Comcast",
    },
    {
      "amount": -300,
      "check_number": "123",
      "date": 2023-08-10T00:00:00.000Z,
      "name": "CHECK 181",
    },
    {
      "amount": -261.46,
      "check_number": "",
      "date": 2023-08-09T00:00:00.000Z,
      "name": "PGANDE",
    },
    {
      "amount": 5000,
      "check_number": "",
      "date": 2023-08-04T00:00:00.000Z,
      "name": "ACME, INC.",
    },
    {
      "amount": -118.28,
      "check_number": "",
      "date": 2023-08-03T00:00:00.000Z,
      "name": "WF Credit Card   AUTO PAY",
    },
    {
      "amount": -220,
      "check_number": "",
      "date": 2023-08-02T00:00:00.000Z,
      "name": "HOA",
    },
    {
      "amount": -500,
      "check_number": "",
      "date": 2023-08-02T00:00:00.000Z,
      "name": "Online Transfer to SAV",
    },
    {
      "amount": -35,
      "check_number": "",
      "date": 2023-08-02T00:00:00.000Z,
      "name": "Online Transfer to CHK",
    },
    {
      "amount": 35,
      "check_number": "",
      "date": 2023-08-02T00:00:00.000Z,
      "name": "Online Transfer from CHK",
    },
    {
      "amount": -2000,
      "check_number": "",
      "date": 2023-08-01T00:00:00.000Z,
      "name": "Rent",
    },
    {
      "amount": 0.01,
      "check_number": "",
      "date": 2023-07-31T00:00:00.000Z,
      "name": "INTEREST PAYMENT",
    },
    {
      "amount": -34.34,
      "check_number": "",
      "date": 2023-07-31T00:00:00.000Z,
      "name": "Uber",
    },
    {
      "amount": -88.11,
      "check_number": "",
      "date": 2023-07-21T00:00:00.000Z,
      "name": "Macdonalds",
    },
    {
      "amount": -82.32,
      "check_number": "",
      "date": 2023-07-21T00:00:00.000Z,
      "name": "ATM WITHDRAW",
    },
    {
      "amount": 5000,
      "check_number": "",
      "date": 2023-07-21T00:00:00.000Z,
      "name": "ACME, INC.",
    },
    {
      "amount": 7.48,
      "check_number": "",
      "date": 2023-07-21T00:00:00.000Z,
      "name": "Uber Refund",
    },
    {
      "amount": -2500,
      "check_number": "",
      "date": 2023-07-20T00:00:00.000Z,
      "name": "Home Fidelity Insurance",
    },
    {
      "amount": -10,
      "check_number": "",
      "date": 2023-07-20T00:00:00.000Z,
      "name": "Online Transfer to CHK .",
    },
    {
      "amount": -3439.22,
      "check_number": "",
      "date": 2023-07-19T00:00:00.000Z,
      "name": "CHASE CREDIT CRD AUTOPAY",
    },
    {
      "amount": 5002.37,
      "check_number": "",
      "date": 2023-07-19T00:00:00.000Z,
      "name": "Online Transfer from SAV",
    },
    {
      "amount": -40,
      "check_number": "",
      "date": 2023-07-17T00:00:00.000Z,
      "name": "VENMO",
    },
    {
      "amount": -300,
      "check_number": "",
      "date": 2023-07-13T00:00:00.000Z,
      "name": "Online Transfer to SAV",
    },
    {
      "amount": -86,
      "check_number": "",
      "date": 2023-07-11T00:00:00.000Z,
      "name": "Comcast",
    },
    {
      "amount": -193.61,
      "check_number": "",
      "date": 2023-07-07T00:00:00.000Z,
      "name": "PGANDE",
    },
    {
      "amount": 5000,
      "check_number": "",
      "date": 2023-07-07T00:00:00.000Z,
      "name": "ACME, INC.",
    },
    {
      "amount": -220,
      "check_number": "",
      "date": 2023-07-05T00:00:00.000Z,
      "name": "HOA",
    },
    {
      "amount": -118.28,
      "check_number": "",
      "date": 2023-07-03T00:00:00.000Z,
      "name": "WF Credit Card   AUTO PAY",
    },
    {
      "amount": -120,
      "check_number": "",
      "date": 2023-07-03T00:00:00.000Z,
      "name": "VENMO",
    },
    {
      "amount": -2000,
      "check_number": "",
      "date": 2023-07-03T00:00:00.000Z,
      "name": "Rent",
    },
  ],
}
`;

exports[`deserializeCSV() > parse-csv-file-headers 1`] = `
[
  {
    "header": "Details",
    "samples": [
      "DEBIT",
      "DEBIT",
      "DEBIT",
      "DEBIT",
      "DEBIT",
    ],
  },
  {
    "header": "Posting Date",
    "samples": [
      "8/21/23",
      "8/21/23",
      "8/21/23",
      "8/21/23",
      "8/21/23",
    ],
  },
  {
    "header": "Description",
    "samples": [
      "PayPal",
      "Uber",
      "CHASE CREDIT CRD AUTOPAY",
      "ATM WITHDRAWAL",
      "Macdonalds",
    ],
  },
  {
    "header": "Amount",
    "samples": [
      "-1400",
      "-22.7",
      "-2332.34",
      "-330",
      "-88.12",
    ],
  },
  {
    "header": "Type",
    "samples": [
      "ACH_DEBIT",
      "DEBIT_CARD",
      "ACH_DEBIT",
      "ATM",
      "ACH_DEBIT",
    ],
  },
  {
    "header": "Balance",
    "samples": [
      "12859.3",
      "14259.3",
      "14282",
      "16614.34",
      "16944.34",
    ],
  },
  {
    "header": "Check or Slip #",
    "samples": [
      "",
      "",
      "",
      "",
      "",
    ],
  },
]
`;

exports[`deserializeCSV() > parse-headers-with-quotes 1`] = `
[
  {
    "header": "Name",
    "samples": [
      "John",
    ],
  },
  {
    "header": "Occupation, Not Separated",
    "samples": [
      "Software Developer",
    ],
  },
  {
    "header": "Age",
    "samples": [
      "30",
    ],
  },
]
`;
